wcfsetup/install/files/lib/system/service/worker/Encryption.class.php

```diff
@@ -0,0 +1,194 @@
+            ECKey::convertPrivateKeyToPEM($privateKey),
+            256
+        );
+        \assert($result);
+        return \str_pad($result, 32, "\x00", STR_PAD_LEFT);
+    }
+
+    private static function addPadding(#[\SensitiveParameter] string $payload, string $contentEncoding): string
+    {
+        $length = \mb_strlen($payload, '8bit');
+        $paddingLength = ServiceWorkerHandler::MAX_PAYLOAD_LENGTH - $length;
+
+        if ($contentEncoding === ServiceWorker::CONTENT_ENCODING_AES128GCM) {
+            return \str_pad($payload . "\x02", $paddingLength + $length, "\x00", STR_PAD_RIGHT);
+        } elseif ($contentEncoding === ServiceWorker::CONTENT_ENCODING_AESGCM) {
+            return \pack('n*', $paddingLength) . \str_pad($payload, $paddingLength + $length, "\x00", STR_PAD_LEFT);
+        } else {
+            throw new \RuntimeException('Unknown content encoding "' . $contentEncoding . '"');
+        }
+    }
+
+    private static function createInfo(string $type, ?string $context, string $contentEncoding)
+    {
+        if ($contentEncoding === ServiceWorker::CONTENT_ENCODING_AESGCM) {
+            \assert($context !== null);
+            \assert(\mb_strlen($context, '8bit') === 135);
+
+            return 'Content-Encoding: ' . $type . "\x00" . Encryption::CURVE_ALGORITHM . $context;
+        } elseif ($contentEncoding === ServiceWorker::CONTENT_ENCODING_AES128GCM) {
+            return 'Content-Encoding: ' . $type . "\x00";
+        } else {
+            throw new \RuntimeException('Unknown content encoding "' . $contentEncoding . '"');
+        }
+    }
+
+    private static function getEncryptionContentCodingHeader(
+        string $contentEncoding,
+        int $length,
+        string $salt,
+        string $publicKey
+    ): string {
+        if ($contentEncoding === ServiceWorker::CONTENT_ENCODING_AES128GCM) {
+            /** {@link https://datatracker.ietf.org/doc/html/rfc8188#section-2.1} */
+            return $salt
+                . \pack('N*', $length)
+                . \pack('C*', \mb_strlen($publicKey, '8bit'))
+                . $publicKey;
+        } elseif ($contentEncoding === ServiceWorker::CONTENT_ENCODING_AESGCM) {
+            return "";
+        } else {
+            throw new \RuntimeException('Unknown content encoding "' . $contentEncoding . '"');
+        }
+    }
+
+    /**
+     * {@link https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-encryption-encoding-00#section-4.2}
+     */
+    private static function createContext(
+        string $clientPublicKey,
+        string $serverPublicKey,
+        string $contentEncoding
+    ): ?string {
+        if ($contentEncoding === ServiceWorker::CONTENT_ENCODING_AES128GCM) {
+            return null;
+        }
+        \assert(\mb_strlen($clientPublicKey, '8bit') === VAPID::PUBLIC_KEY_LENGTH);
+
+        $len = \pack('n', 65);
+
+        return "\x00" . $len . $clientPublicKey . $len . $serverPublicKey;
```

Comment:
Looking at this once more, I believe the following would be more resistant to introducing bugs in a future refactoring, because it actually computes the string length. I've also taken the liberty to merge this into a single pack call, similarly to the session encoder: https://github.com/WoltLab/WCF/blob/e745d4785d7054655d0f036c4b9c6471ef0bd69c/wcfsetup/install/files/lib/system/session/SessionHandler.class.php#L289-L294

```suggestion
        return \pack(
            'CnA*nA*',
            0,
            \strlen($clientPublicKey),
            $clientPublicKey,
            \strlen($serverPublicKey),
            $serverPublicKey,
        );
```